project-koku/koku
# [COST-3847] Create separate processing queues for large customers

## Jira Ticket

[COST-3847](https://issues.redhat.com/browse/COST-3847)

## Description

This change will ...
* Leverage unleash flag with the is_customer_large method
* Route large customers to *_XL queues
* Add new deployments to support the new XL queues
* Capture prometheus metrics for new queues
* Route Masu endpoints to XL queue if a large customer

## Testing

1. Deploy to ephemeral environment
```
bonfire namespace reserve --duration 24h
bonfire deploy hccm --set-template-ref koku=cost-3847-large-customer-queue --set-parameter koku/IMAGE_TAG=pr-4452-1021a7b   --no-remove-resources koku \
  --no-remove-resources presto \
  --no-remove-resources hive-metastore
```
2. Kick off smoke tests `oc apply -f cji-smoke.yml`
```
apiVersion: cloud.redhat.com/v1alpha1
kind: ClowdJobInvocation
metadata:
  name: cost-mgmt-smoke
spec:
  appName: koku
  testing:
    iqe:
      ui:
        enabled: false
      marker: "cost_smoke"
      dynaconfEnvName: "clowder_smoke"
      filter: "test_api"
```
3. Gather results for test when going through regular pipeline and workers ✅ 
`= 194 failed, 4098 passed, 25 skipped, 1200 deselected, 1838 xfailed, 19677 warnings, 3426 errors in 12723.95s (3:32:03) =` and verify no traffic in xl workers
4. Connect to unleash to setup large customer check
`oc port-forward service/env-${NAMESPACE}-featureflags 4242:4242`
5. Add `cost-management.backend.large-customer` to default project as "Operational" feature and set it to "On" for "development" and "production" values (essentially meaning it will always return true.
6. Create new CJI (just change the name) to rerun smokes but now the flow should go through the XL queues and workers
7. Gather results for tests and validate traffic in XL workers ✅ 
`= 87 failed, 4208 passed, 25 skipped, 1200 deselected, 4672 xfailed, 20906 warnings, 589 errors in 16127.07s (4:28:47) =`
verify that while the Orchestrator.prepare() logic does hit the priority queue the download_xl and summary_xl queues and workers handle the traffic and the regular workers had no traffic.
